packages/wallet/notes/action-routing.md

```diff
@@ -0,0 +1,91 @@
+it is important for the protocol to be able to understand which of the two transaction-submitted
+sub-protocols a `TransactionConfirmed` event applies to in this case.
+
+## Proposed Solution
+
+Here's one possible solution to this problem:
+
+1. We replace the `processId` with a `protocolLocator` string.
+2. The `protocolLocator` would look something like `process-123.refute-transaction.attempt-1`.
+3. The first section of the string would be used to identify the process and then the remainder
+   used to route the action within that process.
+4. The information beyond the `processId` is optional. If there's only one path down the protocol
+   tree, then the protocolLocator could just be the `processId`. It's just a place for protocols
+   to add information if required.
+
+## Proposed Implementation
+
+There are many ways to implement this. The key property we need is that each node in the protocol tree
+should have the opportunity to add to the `protocolLocator` for actions emanating from its
+subtree and can then use this information to route any return actions. I propose the following
+approach:
+
+1. Every `protocolState` stores a `protocolLocator` property.
+2. This `protocolLocator` is passed in by the parent. In the case of a top-level protocol
+   the `protocolLocator` would just be the `processId`.
+3. The protocol is responsible for putting the `protocolLocator` onto any outgoing actions/requests
+   and for passing it into any child protocols. It is free to add additional information to
+   the end of the locator before it does this.
+4. When it receives an action it can recover the information added by (1) removing the
+   `protocolLocator` stored in its state and (2) pulling off the next property (if it added one).
+   It can then handle this action or route it to a child as appropriate.
+
+## Alternatives Considered
+
+### Alternative Implementation
+
+One alternative here would be to not store the `protocolLocator` on the `protocolState` and
+instead ensure that all actions are passed back up the protocol tree, allowing the nodes
+above to modify the action itself. This seems nicer in some ways (it seems neat that the
+protocol doesn't need to be aware of its location) but also seems harder to implement:
+we'd need to move back to passing side-effects back from the protocols and come up with a
+strategy for modifying the action creators as we pass them through containers and components.
+
+### Alternative Data Structure
+
+We could make the `processLocator` an object. This object would always have a `processId`
+and optionally have other properties. We might get some type safety benefits from this
+but I can't quite see how this would play out and it might not be worth it.
+
+### Alternative Names
+
+Here are some alternatives for the name of the `processLocator` property:
+`location`, `processLocation`, `route`, `path`, `executionPath`, `protocolLocation`,
```

Comment:
I would probably vote for something like `protocolPath` or `protocolRoute`

Making an analogy to a URL, the `processId` is kind of like the domain name, and the `protocolPath` is like the route that you take from that domain.